### PR TITLE
Add demangling as plugins and libdemangle as external dependency

### LINKED ISSUE #1600: Extract demangling code into the separate library
- [ ] Move the code from `librz/bin/mangling/*` to the separate library `rz-demangle`
- [ ] Make it a Meson subproject in Rizin
- [ ] Move issues and pure demangling tests to this library instead
- [ ] Move `rz-libswift` code either completely or as a subproject of this library as well
- [ ] Move [issues with label "magling"](https://github.com/rizinorg/rizin/labels/mangling) to this new library repository

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Now the demanglers are all handled as external libs (https://github.com/rizinorg/rz-libdemangle).

Also https://github.com/rizinorg/rz-libswift is now a plugin

Closes https://github.com/rizinorg/rizin/issues/1600

### CODE REVIEW

File: librz/bin/bin.c [hunk 1]
```diff
@@ -1293,3 +1293,149 @@ RZ_API const RzBinLdrPlugin *rz_bin_ldrplugin_get(RzBin *bin, const char *name)
 	}
 	return NULL;
 }
+
+static char *bin_demangle_cxx(RzBinFile *bf, const char *symbol, ut64 vaddr) {
```
Comment (XVilka) [inline librz/bin/bin.c:1310]: Please add `rz_return_val_if_fail(symbol)`

Comment (XVilka) [reply librz/bin/bin.c:1310]: Also why `bin_demangle_cxx()` is static while `bin_demangle_rust()` is `RZ_API`?

File: librz/bin/bin.c [hunk 2]
```diff
@@ -1293,3 +1293,149 @@ RZ_API const RzBinLdrPlugin *rz_bin_ldrplugin_get(RzBin *bin, const char *name)
...
+RZ_API char *bin_demangle_rust(RzBinFile *binfile, const char *symbol, ut64 vaddr) {
+	char *str = NULL;
+	if (!(str = bin_demangle_cxx(binfile, symbol, vaddr))) {
+		return str;
+	}
+	free(str);
+	return rz_demangler_rust(symbol);
+}
+
+RZ_API char *rz_bin_demangle(RzBinFile *bf, const char *language, const char *symbol, ut64 vaddr, bool libs) {
```
Comment (XVilka) [inline librz/bin/bin.c]: Please add doxygen, also check `language` for NULL

File: librz/bin/bin.c [hunk 3]
```diff
@@ -1293,3 +1293,149 @@ RZ_API const RzBinLdrPlugin *rz_bin_ldrplugin_get(RzBin *bin, const char *name)
...
+		language = rz_bin_language_to_string(type);
+	}
+	if (!language) {
+		return NULL;
+	}
+	char *demangled = NULL;
+	switch (type) {
+	case RZ_BIN_LANGUAGE_UNKNOWN: return NULL;
+	case RZ_BIN_LANGUAGE_KOTLIN:
+		/* fall-thru */
```
Comment (XVilka) [inline librz/bin/bin.c:1446]: It's `fall-through`

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks much better and cleaner this way 👍 🥇

File: librz/bin/bin.c (hunk 1 above)
Comment (wargio) [reply librz/bin/bin.c:1310]: mistake. it should be static.

File: librz/bin/bin.c (hunk 2 above)
Comment (wargio) [reply librz/bin/bin.c]: there is a check on language below. it is nullable at the end.

File: librz/bin/bin.c (hunk 3 above)
Comment (wargio) [reply librz/bin/bin.c:1446]: it is not a mistake https://developers.redhat.com/blog/2017/03/10/wimplicit-fallthrough-in-gcc-7

File: librz/bin/bin.c (hunk 1 above)
Comment (wargio) [reply librz/bin/bin.c:1310]: > Please add `rz_return_val_if_fail(symbol)`

This can't happen because the symbol value is checked by `rz_bin_demangle`

File: librz/bin/bin.c (hunk 3 above)
Comment (XVilka) [reply librz/bin/bin.c:1446]: It's a bad language. What about `/* Falls through. */`?

**Review by XVilka — APPROVED** [review]

File: librz/bin/bin.c (hunk 3 above)
Comment (wargio) [reply librz/bin/bin.c:1446]: for me is ok. i can make the change in another PR because there are multiple `/* fall-thru */` in the codebase

Comment (wargio) [reply librz/bin/bin.c:1446]: https://github.com/rizinorg/rizin/issues/1674

**Comment by XVilka** [comment]
@wargio please rebase again to fix the Windows builds.

File: librz/bin/bin.c [hunk 4]
```diff
@@ -1295,3 +1295,159 @@ RZ_API const RzBinLdrPlugin *rz_bin_ldrplugin_get(RzBin *bin, const char *name)
...
+		return out;
+	}
+	char *sign = (char *)strchr(out, '(');
+	if (!sign) {
+		return out;
+	}
+
+	char *str = out;
+	char *ptr = NULL;
+	char *method_name = NULL;
```
Comment (ret2libc) [inline librz/bin/bin.c:1322]: is it reasonable to have this method extraction thing moved into libdemangler? I expect it's a common operation across languages to extract the method name of a demangled string, right?

File: librz/bin/bin_language.c [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+/**
+ * \brief Tries to detect which language is used in the binary based on symbols and libraries
+ *
+ * Currently this method can detect the language only from bins that are either ELF, PE,
+ * Mach-O, Java Class and Dex.
+ * 
+ * The current supported languages are: c, cxx, dart, dlang, go, groovy, java, kotlin, msvc,
+ * objc, rust, swift.
+ */
+RZ_API RzBinLanguage rz_bin_language_detect(RzBinFile *binfile) {
```
Comment (ret2libc) [inline librz/bin/bin_language.c:67]: do we have at least one test for each case? c++, rust, dart, etc. ?

File: librz/core/cmd_descs/cmd_info.yaml [hunk 6]
```diff
@@ -137,21 +137,34 @@ commands:
...
+          - name: lang
+            type: RZ_CMD_ARG_TYPE_CHOICES
+            choices:
+              - c++
+              - java
+              - objc
+              - swift
+              - dlang
+              - msvc
+              - rust
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_info.yaml:155]: Maybe later it's worth to introduce a new ARG_TYPE which is autocompleted at runtime instead of having this manual list

File: librz/core/cmd_info.c [hunk 7]
```diff
@@ -594,8 +570,51 @@ RZ_IPI RzCmdStatus rz_cmd_info_pdb_download_handler(RzCore *core, int argc, cons
 	return RZ_CMD_STATUS_OK;
 }
 
+static bool print_demangler_info(const RzDemanglerPlugin *plugin, void *user) {
+	if (!user) {
+		rz_cons_printf("%-6s %-8s %s\n", plugin->language, plugin->license, plugin->author);
+		return true;
+	}
```
Comment (ret2libc) [inline librz/core/cmd_info.c:563]: split this in a separate function.

File: meson_options.txt [hunk 8]
```diff
@@ -33,6 +33,8 @@ option('use_sys_openssl', type: 'feature', value: 'disabled')
 option('use_sys_tree_sitter', type: 'feature', value: 'disabled')
 option('use_sys_libuv', type: 'feature', value: 'auto', description: 'Whether to force, suggest or not use at all the system version of libuv. If system version is not found, one is built statically, unless use_libuv is false.')
 option('use_libuv', type: 'boolean', value: true, description: 'If true, libuv is used to handle remote features')
+option('use_swift_demangler', type: 'boolean', value: true, description: 'If false, disables the swift demangler')
+option('use_gpl', type: 'boolean', value: true, description: 'Set to false when you want to disable gpl code')
```
Comment (ret2libc) [inline meson_options.txt:37]: I'd add "not completely supported yet" because I think there's other code that is GPL which is not disabled if you set this to false, right?

File: librz/core/cmd_descs/cmd_info.yaml (hunk 6 above)
Comment (wargio) [reply librz/core/cmd_descs/cmd_info.yaml:155]: that is a great idea

File: librz/bin/bin.c (hunk 4 above)
Comment (wargio) [reply librz/bin/bin.c:1322]: this is more a hack of the old code that adds classes and methods to the bin based on the demangled output. i would keep this here and then fix this by having this logic within bin by looking for this stuff.

File: librz/bin/bin_language.c (hunk 5 above)
Comment (wargio) [reply librz/bin/bin_language.c:67]: in tests, yes with the exclusion of dlang and dart.

File: meson_options.txt (hunk 8 above)
Comment (wargio) [reply meson_options.txt:37]: if you mean in rizin codebase, yes, but if it is referred to libdemangler, no. it is supported and not built when this flag is enabled.

Comment (ret2libc) [reply meson_options.txt:37]: i mean in rizin codebase. This is rizin meson file, so one may think that it applies to the whole rizin.

Comment (wargio) [reply meson_options.txt:37]: then no. but i could find all GPL code and add the `#if USE_GPL` in another commit.

Comment (ret2libc) [reply meson_options.txt:37]: i think it's a separate thing. Maybe let's create an issue and do it separately after this if you want.

File: librz/bin/bin.c (hunk 4 above)
Comment (ret2libc) [reply librz/bin/bin.c:1322]: ok yeah if it's all moved/refactored code that's fine.

File: meson_options.txt (hunk 8 above)
Comment (wargio) [reply meson_options.txt:37]: ok, i have added the comment. i will make another PR with all GPL guards once this is merged.

Comment (XVilka) [reply meson_options.txt:37]: There are already issues for that:
- https://github.com/rizinorg/rizin/issues/70
- https://github.com/rizinorg/rizin/issues/209

**Review by XVilka — APPROVED** [review]

File: librz/main/rizin.c [hunk 9]
```diff
@@ -389,6 +391,7 @@ RZ_API int rz_main_rizin(int argc, const char **argv) {
 	rz_signal_sigmask(SIG_BLOCK, &sigBlockMask, NULL);
 #endif
 
+	rz_demangler_plugin_init();
```
Comment (ret2libc) [inline librz/main/rizin.c]: Why is this needed? I think it should not be here... maybe somewhere inside RzCore? Please look at how all the other plugins are initialized and do the same, this feels wrong.

**Review by ret2libc — CHANGES_REQUESTED** [review]
I think most things look good, apart from the following:
Most `librz/main` binaries should not have to care about rz_demanglaer_plugin_init/fini IMHO. Maybe only rz-bin needs it.

- [ ] TODO: create issue to create new RZ_ARG_TYPE_LANG with autocompletion for `iD`

File: librz/main/rizin.c (hunk 9 above)
Comment (wargio) [reply librz/main/rizin.c]: no, i disagree.
it should be initialized by the tool only once and i want to avoid to implement ref counting for this.
i do not want to place this into RzCore because rz-bin needs it since it provides demangling and you can create n-th instances of the RzCore.
Also it is required to have dynamically allocated lists where plugins of this type are registered when loading shared libs.

Comment (ret2libc) [reply librz/main/rizin.c]: I'm not convinced. Look at how other plugins do their initialization. Look at config.h.in and maybe let's add a `RZ_DEMANGLER_STATIC_PLUGINS` there with the list of plugins you need to build. You can easily define them in meson.build instead of at runtime.

Comment (wargio) [reply librz/main/rizin.c]: it does not change the fact that they cannot be initialized by RzCore

Comment (ret2libc) [reply librz/main/rizin.c]: You do not need to initialize/finalize anything if you just move the list in RZ_DEMANGLER_STATIC_PLUGINS, like what's done for all the other plugins types. It is no different than a RzBin plugin, needed by rz-bin.

Comment (wargio) [reply librz/main/rizin.c]: no. all the plugins are still initialized when you call the `rz_<something>_new()`

File: librz/demangler/demangler.c [hunk 10]
```diff
@@ -0,0 +1,204 @@
+// SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
+// SPDX-FileCopyrightText: 2021 deroad <wargio@libero.it>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_demangler.h>
+#include <rz_util.h>
+#include <rz_list.h>
+#include <rz_libdemangle.h>
+
+static RzList *demangler_plugins = NULL;
```
Comment (ret2libc) [inline librz/demangler/demangler.c]: Just to wrap up what was discussed on mattermost:
1) no global
2) use RzDemangler object to keep track of demangler_plugins (and whatever can come later)
3) use `RZ_DEMANGLER_STATIC_PLUGINS` in config.h.in and initialize those plugins in `rz_demangler_new`

**Review by ret2libc — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Comment by wargio** [comment]
requires a version bump once https://github.com/rizinorg/rz-libdemangle/pull/10 is merged
**edit**: merged.

File: librz/demangler/demangler.c [hunk 11]
```diff
@@ -0,0 +1,184 @@
+// SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
+// SPDX-FileCopyrightText: 2021 deroad <wargio@libero.it>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_demangler.h>
+#include <rz_util.h>
+#include <rz_libdemangle.h>
+
+#include "demangler_plugins.c"
```
Comment (ret2libc) [inline librz/demangler/demangler.c]: let's avoid this inclusion of c files in new files if not strictly necessary. Just put the whole content here, I think it's fine.

File: librz/include/rz_demangler.h [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+#define rz_demangler_plugin_demangle(x, y) ((x) && RZ_STR_ISNOTEMPTY(y) ? (x)->demangle(y) : NULL)
+
+RZ_API RZ_OWN char *rz_demangler_java(RZ_NULLABLE const char *symbol);
+RZ_API RZ_OWN char *rz_demangler_cxx(RZ_NONNULL const char *symbol);
+RZ_API RZ_OWN char *rz_demangler_objc(RZ_NONNULL const char *symbol);
+RZ_API RZ_OWN char *rz_demangler_rust(RZ_NONNULL const char *symbol);
+RZ_API RZ_OWN char *rz_demangler_msvc(RZ_NONNULL const char *symbol);
+
+RZ_API RZ_OWN RzDemangler *rz_demangler_new();
```
Comment (ret2libc) [inline librz/include/rz_demangler.h]:
```suggestion
RZ_API RZ_OWN RzDemangler *rz_demangler_new(void);
```
?? i don't remember if it's the same or not..

File: librz/main/rz-gg.c [hunk 13]
```diff
@@ -6,6 +6,7 @@
 #include <rz_main.h>
 #include <rz_util/rz_print.h>
 #include <rz_util.h>
+#include <rz_demangler.h>
```
Comment (ret2libc) [inline librz/main/rz-gg.c]: not necessary?

File: librz/main/rz-find.c [hunk 14]
```diff
@@ -7,6 +7,7 @@
 #include <rz_main.h>
 #include <rz_types.h>
 #include <rz_search.h>
+#include <rz_demangler.h>
```
Comment (ret2libc) [inline librz/main/rz-find.c]: not necessary?

File: librz/main/rz-diff.c [hunk 15]
```diff
@@ -6,6 +6,7 @@
 #include <rz_bin.h>
 #include <rz_diff.h>
 #include <rz_util.h>
+#include <rz_demangler.h>
```
Comment (ret2libc) [inline librz/main/rz-diff.c]: not necessary?

File: librz/main/rz-sign.c [hunk 16]
```diff
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 #include <rz_main.h>
 #include <rz_core.h>
+#include <rz_demangler.h>
```
Comment (ret2libc) [inline librz/main/rz-sign.c]: not necessary?

**Review by ret2libc — COMMENTED** [review]
LGTM

File: librz/include/rz_demangler.h (hunk 12 above)
Comment (wargio) [reply librz/include/rz_demangler.h]: yes, it is the same. i can add that.

**Review by ret2libc — APPROVED** [review]
